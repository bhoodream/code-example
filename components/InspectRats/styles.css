.InspectRats {
    display: flex;
    flex-direction: column;
    height: 100vh;
}

@media (min-width: 550px) {
    .InspectRats {
        flex-direction: row;
    }
}
