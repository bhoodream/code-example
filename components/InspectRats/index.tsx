import React, { Component } from 'react';
import axios from 'axios';

import Rat from '../Rat/index';

import { RatName, RatData, RatDataMap } from '../../types';
import { EMPTY_RAT_OPTION } from '../../Const';

import getKeyValue from '../../utils/types/getKeyValue';

import './styles.css';

type InspectRatsProps = {
    rats: string[];
};

interface InspectRatsState {
    currentRatName: RatName;
    ratDataMap: RatDataMap;
    loading: boolean;
}

class InspectRats extends Component<InspectRatsProps, InspectRatsState> {
    state = {
        currentRatName: EMPTY_RAT_OPTION.val,
        loading: false,
        ratDataMap: {},
    };

    static defaultProps = {
        rats: [],
    };

    loadRatData(name: RatName) {
        if (!name || this.getRatData(name)) {
            return this.setState({ currentRatName: name });
        }

        this.setState({ currentRatName: name, loading: true }, () => {
            axios
                .get(`http://localhost:7421/rat/${name}`)
                .then(({ data }) => this.onRatDataLoaded(name, data))
                .catch((e) => console.error(e));
        });
    }

    onRatDataLoaded(name: RatName, data: RatData) {
        this.setState((state: InspectRatsState) => {
            loading: false,
            ratDataMap: {
                ...state.ratDataMap,
                [name]: data,
            },
        });
    }

    getRatData(name: RatName) {
        return getKeyValue<RatDataMap, keyof RatDataMap>(
            this.state.ratDataMap,
            name
        );
    }

    render() {
        const { rats } = this.props;
        const { currentRatName, loading } = this.state;
        const currentRatData: RatData | undefined = this.getRatData(
            currentRatName
        );

        return (
            <div className={'InspectRats'}>
                <div className={'InspectRat__select'}>
                    <select
                        disabled={loading}
                        onChange={(e) => this.loadRatData(e.target.value)}
                        defaultValue={EMPTY_RAT_OPTION.val}
                    >
                        <option
                            key={EMPTY_RAT_OPTION.label}
                            value={EMPTY_RAT_OPTION.val}
                        >
                            {EMPTY_RAT_OPTION.label}
                        </option>
                        {rats.map((r: string) => (
                            <option key={r} value={r}>
                                {r}
                            </option>
                        ))}
                    </select>
                </div>
                {loading && 'Loading...'}
                {currentRatData && (
                    <div className={'InspectRat__data'}>
                        <Rat {...currentRatData} />
                    </div>
                )}
            </div>
        );
    }
}

export default InspectRats;
