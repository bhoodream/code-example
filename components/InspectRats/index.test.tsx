import React from 'react';
import renderer from 'react-test-renderer';
import { render, act, fireEvent, waitForElement } from '@testing-library/react';

import createTestServer from '../../utils/tests/createTestServer';

import InspectRats from './index';

const RATS = ['rat1', 'rat2', 'rat3'];
const server = createTestServer();

describe('InspectRats', () => {
    it('renders without crashing', () => {
        render(<InspectRats rats={RATS} />);
    });

    it('get rat data', async () => {
        server.get('http://localhost:7421/rat/rat2', () => ({
            width: 3,
            height: 4,
            nickname: 'nick of rat2',
        }));

        const { getByRole, getByText, queryByText } = render(
            <InspectRats rats={RATS} />
        );
        const select = getByRole('combobox');

        act(() => {
            fireEvent.change(select, { target: { value: 'rat2' } });
        });

        const selectedOption = getByText('rat2') as HTMLOptionElement;

        expect(select.hasAttribute('disabled')).toBe(true);
        expect(selectedOption.selected).toBe(true);
        expect(getByText('Loading...')).toBeDefined();

        await waitForElement(() => getByText('width: 3'));
        await waitForElement(() => getByText('height: 4'));
        await waitForElement(() => getByText('nickname: nick of rat2'));

        expect(queryByText('Loading...')).toBeNull();
    });

    it('snapshots', () => {
        const tree = renderer.create(<InspectRats rats={RATS} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
