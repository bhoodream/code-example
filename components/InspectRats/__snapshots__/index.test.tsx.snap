// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectRats snapshots 1`] = `
<div
  className="InspectRats"
>
  <div
    className="InspectRat__select"
  >
    <select
      defaultValue=""
      disabled={false}
      onChange={[Function]}
    >
      <option
        value=""
      >
        No Rat
      </option>
      <option
        value="rat1"
      >
        rat1
      </option>
      <option
        value="rat2"
      >
        rat2
      </option>
      <option
        value="rat3"
      >
        rat3
      </option>
    </select>
  </div>
</div>
`;
